### PR TITLE
Keyboard test fixes v2

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

https://github.com/dotnet/maui/pull/14388 was merged a bit early. I should have left it as draft, we were still seeing occasional failures.

### CODE REVIEW

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs [hunk 1]
```diff
@@ -176,10 +174,20 @@ public static async Task HideKeyboardForView(this AView view, int timeout = 1000
...
-					throw new Exception("HideKeyboard operation failed");
-
-				await view.WaitForKeyboardToHide(timeout);
+				int retry = 2;
+				for (int i = 0; i < retry; i++)
+				{
+					KeyboardManager.HideKeyboard(view);
+					try
+					{
+						await view.WaitForKeyboardToHide(timeout);
```
Comment (mandel-macaque) [inline src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs:183]: Always retry with the same timeout? Shall we increase in every iteration?

Comment (PureWeen) [reply src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs]: reworked to use are `wait` method that already has backoff behavior.

File: eng/devices/android.cake [hunk 2]
```diff
@@ -213,8 +213,9 @@ Task("Test")
 	Information("Test App Package Name: {0}", TEST_APP_PACKAGE_NAME);
 	Information("Test App Instrumentation: {0}", TEST_APP_INSTRUMENTATION);
 	Information("Test Results Directory: {0}", TEST_RESULTS);
-
-	CleanDirectories(TEST_RESULTS);
+	
+	if (IsCIBuild())
+		CleanDirectories(TEST_RESULTS);
```
Comment (PureWeen) [inline eng/devices/android.cake:218]: Is this ok @rmarinho? Will there be results left over from other CI runs that we need to clear out?

File: eng/devices/android.cake [hunk 3]
```diff
@@ -213,8 +213,15 @@ Task("Test")
...
-
-	CleanDirectories(TEST_RESULTS);
+	
+	if (!IsCIBuild())
+		CleanDirectories(TEST_RESULTS);
+	else
+	{
+		// Because we retry on CI we don't want to delete the previous failures
+		// We want to publish those files for reference
+		DeleteFiles(Directory(TEST_RESULTS).Path.Combine("*.*").FullPath);
```
Comment (PureWeen) [inline eng/devices/android.cake:223]: Is this ok @rmarinho? Will there be results left over from other CI runs that we need to clear out?

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs [hunk 4]
```diff
@@ -155,8 +155,23 @@ public static async Task ShowKeyboardForView(this AView view, int timeout = 1000
...
 				KeyboardManager.ShowKeyboard(view);
-				await view.WaitForKeyboardToShow(timeout);
+				await Task.Yield();
+
+				bool result = await Wait(() =>
+				{
+					bool isVisible = KeyboardManager.IsSoftKeyboardVisible(view);
+
+					if (!isVisible)
+						KeyboardManager.ShowKeyboard(view);
```
Comment (rachelkang) [inline src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs:167]: Do we need to reset isVisible here as well?

If isVisible was false before, and then we show the keyboard, then the keyboard becomes visible, so I assume isVisible should be updated before returning it?

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs [hunk 5]
```diff
@@ -174,20 +189,21 @@ public static async Task HideKeyboardForView(this AView view, int timeout = 1000
...
-						if (i >= retry)
-							throw;
-					}
-				}
+					bool isHidden = !KeyboardManager.IsSoftKeyboardVisible(view);
+
+					if (!isHidden)
+						KeyboardManager.HideKeyboard(view);
+
+					return isHidden;
```
Comment (rachelkang) [inline src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs:201]: same question as above

**Review by rachelkang — CHANGES_REQUESTED** [review]
am seeing intermittent failures locally with the Modal keyboard and Adjust/Pan keyboard test :(

but otherwise, changes here LGTM!

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs (hunk 4 above)
Comment (PureWeen) [reply src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs:165]: changed

File: src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs (hunk 5 above)
Comment (PureWeen) [reply src/TestUtils/src/DeviceTests/AssertionExtensions.Android.cs]: changed

File: src/Controls/tests/DeviceTests/Elements/Application/ApplicationTests.Android.cs [hunk 6]
```diff
@@ -59,24 +86,7 @@ public async Task SoftInputModeSetOnApplicationPropagatesToWindowHandlers()
 
 						handlers.Insert(0, app.Window.Handler);
 
-						// Validate that the Soft Input initializes to AdjustPan
-						Assert.Equal(ASoftInput.AdjustPan, windowHandler.LastASoftInputSet);
```
Comment (rachelkang) [inline src/Controls/tests/DeviceTests/Elements/Application/ApplicationTests.Android.cs:63]: Did you mean to remove this Assert? (AFAICT everything else except for this line was moved)

Comment (PureWeen) [reply src/Controls/tests/DeviceTests/Elements/Application/ApplicationTests.Android.cs:63]: Yea, this is already being validated by the other test

https://github.com/dotnet/maui/pull/14397/files#diff-2f5fa116e225dc8e883f21962e576c8b13855ed2a873633b51b1d2ebebdfcf9bR24

So, it's not necessary to validate here as well.

**Review by rachelkang — APPROVED** [review]

File: eng/devices/android.cake (hunk 3 above)
Comment (rmarinho) [reply eng/devices/android.cake:223]: Hum, the macos images are not clear, so in a next test run there could be files there?

I wonder if we just need a step to clear this on the yaml timeline when the pipeline starts

**Pull request merged**